frontend/app/src/components/display/amount/AmountDisplay.vue

```diff
@@ -0,0 +1,404 @@
+);
+
+const { exchangeRate, assetPrice, getHistoricPrice } = useBalancePricesStore();
+
+const {
+  abbreviateNumber,
+  thousandSeparator,
+  decimalSeparator,
+  currencyLocation,
+  amountRoundingMode,
+  valueRoundingMode
+} = storeToRefs(useFrontendSettingsStore());
+
+const { isAssetPriceInCurrentCurrency } = useBalancePricesStore();
+
+const isCurrentCurrency = isAssetPriceInCurrentCurrency(priceAsset);
+
+const { findCurrency } = useCurrencies();
+
+const historicExchangeRate = asyncComputed(
+  async () => {
+    if (!isDefined(sourceCurrency)) {
+      return One.negated();
+    }
+    return await getHistoricPrice({
+      fromAsset: get(sourceCurrency),
+      toAsset: get(currentCurrency),
+      timestamp: get(timestamp)
+    });
+  },
+  One.negated(),
+  {
+    lazy: true,
+    evaluating
+  }
+);
+
+const latestFiatValue: ComputedRef<BigNumber> = computed(() => {
+  const currentValue = get(value);
+  const to = get(currentCurrency);
+  const from = get(sourceCurrency);
+
+  if (!from || to === from) {
+    return currentValue;
+  }
+  const multiplierRate = to === CURRENCY_USD ? One : get(exchangeRate(to));
+  const dividerRate = from === CURRENCY_USD ? One : get(exchangeRate(from));
+
+  if (!multiplierRate || !dividerRate) {
+    return currentValue;
+  }
+  return currentValue.multipliedBy(multiplierRate).dividedBy(dividerRate);
+});
+
+const internalValue: ComputedRef<BigNumber> = computed(() => {
+  // If there is no `sourceCurrency`, it means that no fiat currency, or the unit is asset not fiat, hence we should just show the `value` passed.
+  // If `forceCurrency` is true, we should also just return the value.
+  if (!isDefined(sourceCurrency) || get(forceCurrency)) {
+    return get(value);
+  }
+
+  const sourceCurrencyVal = get(sourceCurrency)!;
+  const currentCurrencyVal = get(currentCurrency);
+  const priceAssetVal = get(priceAsset);
+  const isCurrentCurrencyVal = get(isCurrentCurrency);
+
+  // If `priceAsset` is defined, it means we will not use value from `value`, but calculate it ourselves from the price of `priceAsset`
+  if (priceAssetVal) {
+    if (priceAssetVal === currentCurrencyVal) {
+      return get(amount);
+    }
+
+    // If `isCurrentCurrency` is true, we should calculate the value by `amount * priceOfAsset`
+    if (isCurrentCurrencyVal && isDefined(priceOfAsset)) {
+      const priceOfAssetVal = get(priceOfAsset) || get(assetPrice(priceAsset));
+      return get(amount).multipliedBy(priceOfAssetVal);
+    }
+  }
+
+  // If `sourceCurrency` and `currentCurrency` is not equal, we should convert the value
+  if (sourceCurrencyVal !== currentCurrencyVal) {
+    let calculatedValue = get(latestFiatValue);
+
+    if (get(timestamp) > 0) {
+      const historicRate = get(historicExchangeRate);
+      if (historicRate.isPositive()) {
+        calculatedValue = get(value).multipliedBy(historicRate);
+      } else {
+        calculatedValue = Zero;
+      }
+    }
+
+    return calculatedValue;
+  }
```

Comment:
I guess we can return early here.

```
// If `sourceCurrency` and `currentCurrency` is not equal, we should convert the value
  if (sourceCurrencyVal !== currentCurrencyVal) {
    if (get(timestamp) > 0) {
      const historicRate = get(historicExchangeRate);
      if (historicRate.isPositive()) {
        return get(value).multipliedBy(historicRate);
      }
      return Zero;
    }

    return get(latestFiatValue);
  }
```

Reply:
This is intentionally, to reduce the number of return statements in the function.